### PR TITLE
Add support for Backend Functions and add use_gpu arg for Setup

### PR DESCRIPTION
**Pull request by chaithyagr**
This PR Works on first part of #61 :

* Configuration check, ie, have a variable that notifies if GPU compute is even possible, this sets the stage for many other computations.
* Find where data resides based on input
* Move data to a specific target (CPU / GPU)
* Obtain the module for coding, ie np / cp

Currently I have added a parameter called use_gpu, which is by default False, that would make the codes function as earlier. This will ensure backward compatibility.

Adding Zac and Sam to review. In my opinion, as this doesnt break things, we can merge this to master, but we can discuss about this in detail in this PR and any potential issues of doing this.

### CODE REVIEW

File: modopt/opt/algorithms.py [hunk 1]
```diff
@@ -502,8 +519,8 @@ def __init__(self, x, grad, prox, cost='auto', beta_param=1.0,
 
         # Set the initial variable values
         self._check_input_data(x)
-        self._x_old = np.copy(x)
-        self._z_old = np.copy(x)
+        self._x_old = self.copy_data(x)
+        self._z_old = self.copy_data(x)
```
Comment (chaithyagr) [inline modopt/opt/algorithms.py]: As this issue focusses only on ForwardBackward, I have added move data to GPU function only here.
Again, if use_gpu is False, it is still np.copy which is done.

File: .idea/.gitignore [hunk 2]
```diff
@@ -0,0 +1,2 @@
+# Default ignored files
+/workspace.xml
```
Comment (sfarrens) [inline .idea/.gitignore]: Missing newline at end of file

**Comment by sfarrens** [comment]
@chaithyagr Everything looks good to me, I would just like to run some tests before merging.

File: .idea/.gitignore (hunk 2 above)
Comment (chaithyagr) [reply .idea/.gitignore]: Ahh, this is not needed, this is an artifact of my IDE, will remove this.

**Comment by chaithyagr** [comment]
By tests you mean GPU tests? Well, this is just the first set of codes, and I dont see any GPU codes working unless all conditions are met (ie all operations MUST take and return GPU data)
However, I have already written such functions and optimizations for other projects. 

The current codes just adds sufficient backend support to kick start the process. ie Part 1 in https://github.com/CEA-COSMIC/ModOpt/issues/61#issuecomment-559848143

Also, I have ensured that we will not break legacy in any case and right now it is experimental. 
Due to this, I was requesting for a merge not only to GPU branch but also mainline if possible. (So that any new issues that I or the tests may have missed can be directly dealt with)

**Comment by sfarrens** [comment]
@chaithyagr I was planning to simply run some tests on a GPU with and without cupy installed to make sure all the exceptions hold etc.

**Comment by chaithyagr** [comment]
@sfarrens , just adding that I dont think things will work yet, it is only after all operators working out of GPU that we can expect this.

**Comment by chaithyagr** [comment]
I am just rebasing this branch to keep it updated. @sfarrens shall we close this PR while we wait for best approach?
This might prevent unnecessary emails about my updates to this branch.

**Pull request closed without merging**